File: Week 02/id_063/Java/LeetCode_429_063.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};

/*
思路
进行DFS，各层节点被遍历到的时候将自己的数值加入到其层数对应的List中，
节点访问顺序并没有要求要按照其层次顺序进行，没必要进行BFS
 */

class Solution {
    private List<List<Integer>> dfs(List<List<Integer>> result, Node curNode, int level) {
        if (curNode == null) {
            return result;
        }

        if (result.size() < level+1) {
            while (result.size() < level+1) {
                result.add(new LinkedList<>());
            }
        }

        result.get(level).add(curNode.val);
        if (curNode.children != null) {
            for (Node subNode : curNode.children) {
                dfs(result, subNode, level+1);
            }
        }

        return result;
    }


    public List<List<Integer>> levelOrder(Node root) {
        return dfs(new ArrayList<List<Integer>>(1000), root, 0);
    }
}

File: Week 02/id_063/C++/LeetCode_429_063.cpp
/*
思路
进行DFS，各层节点被遍历到的时候将自己的数值加入到其层数对应的List中，
节点访问顺序并没有要求要按照其层次顺序进行，没必要进行BFS
 */


#include <vector>
using namespace std;

// Definition for a Node.
/*
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
 */

class Solution {
public:

    void dfs(vector<vector<int>>& result, Node* p_cur_node, int level) {
        if (p_cur_node == nullptr) {
            return;
        }

        while (result.size() < level + 1) {
            result.emplace_back();
        }

        result[level].push_back(p_cur_node->val);
        for (Node* p_sub_node : p_cur_node->children) {
            dfs(result, p_sub_node, level + 1);
        }
    }

    vector<vector<int>> levelOrder(Node* root) {
        vector<vector<int>> result;
        dfs(result, root, 0);

        return result;
    }
};